### ISSUE TITLE
[RFC] Introduce LightningClosure

### ISSUE THREAD

**Issue by tchaton** [PyTorchLightning/pytorch-lightning#9129]
## Proposed refactoring or deprecation

Work and description copy / pasted from @awaelchli 

### Motivation: 

Depending on how involved a loop customization is, sooner or later a user has to deal with closures at some point. This is quite difficult in Lightning currently due to a high entanglement with gradient accumulation, manual backward, precision etc. By extracting out a high level api for building closures, we enable customization while providing a clean implementation for manual optimization etc. 

### Proposed Solution

A Lightning Closure with the following responsibilities:
- takes three functions as input (step, backward, zero grad) and calls them in the right order.
- caches the output of the step (training_step) function, to be accessed outside once the closure has run
- includes profiling on top of the three functions

```python
class LightningClosure(Closure):
    def __init__(
        self,
        step_fn: Callable[[], dict],
        backward_fn: Optional[Callable[[Tensor], Tensor]] = None,
        zero_grad_fn: Optional[Callable[[], None]] = None,
        profiler: Optional[BaseProfiler] = None,
    ):
```

A Lightning Closure gets created inside our training loop like so (pseudo code):

```py
closure = LightningClosure(step_fn=self._training_step, backward_fn=self._backward, zero_grad_fn=self._zero_grad)
```

and then later on it gets passed to the optimizer after which we can access results:

```py
optimizer.step(closure)
result = closure.result
```

With the help of the closure abstraction, it is now very easy to implement manual optimization by setting

```py
closure = LightningClosure(step_fn=self._manual_training_step, backward_fn=None, zero_grad_fn=None)
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

**Comment by ananthsub** [comment]
> no. the closure is strictly for implementing loops. there will be no additional abstractions in users LightningModule code! they can create a closure just with a regular python function if they need to.

given this, shouldn't we be able to create closures inside the loop with a regular python function too?

**Comment by awaelchli** [comment]
> given this, shouldn't we be able to create closures inside the loop with a regular python function too?

of course, that's how it was done prior to the PR, but this doesn't help us get any closer to loop customization unless the user wants to duplicate tons of code from our loops. an example is writing new loops for new flavors of the training_step #8643 .

why is there such resistance against internal abstraction? 

our approach offers:

- code is self-documenting (versus having lengthy docstrings in god classes or not documentation at all due to laziness)
- splitting responsibility (versus having god classes = higher maintenance cost)
- less code duplication (versus having code duplication and higher maintenance cost)
